services/azure-devops/vso-redirect.service.js

```diff
@@ -0,0 +1,30 @@
+'use strict'
+
+const { redirector } = require('..')
+
+module.exports = [
+  redirector({
+    category: 'build',
+    route: {
+      base: 'vso/build',
+      format: '([^/]+)/([^/]+)/([^/]+)(?:/(.+))?',
+      capture: ['organization', 'projectId', 'definitionId', 'branch'],
+    },
+    target: ({ organization, projectId, definitionId, branch }) => {
+      let path = `/azure-devops/build/${organization}/${projectId}/${definitionId}`
```

Comment:
It would be cool if we could grab this value from the route defined in the corresponding service (azure-devops-build service in this case) to having these defined in multiple places. 

If that's not possible/feasible, is there anything we could do to help avoid route updates being made in a service but forgotten here? Maybe another danger check?

Reply:
I started to think that maybe this should use path-to-regexp compile. After thinking about it for a while, I concluded it wasn't worth it for two reasons:

1. With few exceptions, we don't make breaking changes in routes. When we change routes, we are careful to keep the old ones working. So if the target route changes, it's likely we will have kept the old target route working, too.
2. There are not going to be a massive amount many of these redirects, so this feels less important to get perfect compared to other parts of the codebase like the core services and badge helper functions (both of which could use attention). It doesn't seem essential to make this totally DRY.

Reply:
> I concluded it wasn't worth it for two reasons

I'm good with that 

> 1. When we change routes, we are careful to keep the old ones working 

That's the exact use case I was thinking through, and wanting to make sure that when/if that happens, that route updates would have to be made here to prevent breaking users leveraging the old route. 

> 2. There are not going to be a massive amount many of these redirects, so this feels less important to get perfect compared to other parts of the codebase...

Also agreed, doesn't need to be a blocker. I've had to do a few of these pattern-based old/new routes while refactoring services so just wanted to pose the question to avoid any potential future route changes being missed on redirects

Reply:
> > 1. When we change routes, we are careful to keep the old ones working
>
> That's the exact use case I was thinking through, and wanting to make sure that when/if that happens, that route updates would have to be made here to prevent breaking users leveraging the old route.

If we keep the old one working, updating this is optional.

1. /vso/foo/bar redirects to /azure-devops/foo/bar
2. /azure-devops/foo/bar is changed to /azure-devops/build/foo/bar.
3. To provide backward compatibility with badge URLs in the wild, a new redirect is added, redirecting /azure-devops/foo/bar to /azure-devops/build/foo/bar
4. /vso/foo/bar could be updated too. That would eliminate an extra redirect. However if it isn't, it will still work because of 3.

Reply:
Step 3 is something the user would have to add to the redirect service as part of updating the route (step 1) right?

Reply:
Hmm, no I don't think so. Step 3 is something that would happen in conjunction with Step 2, by the Shields developer who decides to change the route again.

Reply:
Oops, bad wording on my part. I meant user in the context of a developer sorry.